### PR TITLE
Monkey: Replace Table-Walk With Table-Parkour

### PR DESCRIPTION
**Pull request by MarkNais**
[REWORK] [FEEDBACK]
## About the PR 
Monkeys can no loner arbitrarily walk through tables. Instead, clicking a table with an empty hand, or "bumping" into them (walk into them) starts an action bar for vaulting over the table(s).
This extends @Tenryumi railing code. The monkey will attempt to vault any length of tables, and will succeed if the length is less than 7, and the destination is passable. The action-bar duration scales with tables traveled (currently, 1 second per-table),

This gif is supposed to embed, but the code ain't working for me.
Note: it is out-of-date. Captured before table-bumping caused the actionbar to start.
https://cdn.discordapp.com/attachments/383743035894267905/733902417304879104/G4fFFlJwW1.gif

## Why's this needed? 
Monkeys hiding under tables *and* interacting with things around them is, in my opinion, bad.
I think this approach, of bringing them similar to humans, but still able to cross tables, is better than the current behavior.

One idea I had but did not implement: short stun on landing. If a monkey is parkouring into your area, you'd have just enough time to grab-table-shove them back.

## Changelog

```
(u)MarkNstein:
(*)Monkey-table interaction changed: Tables cannot be walked through, but instead can be vaulted past.
```

### CODE REVIEW

**Comment by TDHooligan** [comment]
Monkeys hopping over bars and tables like they're nothing can be really distracting and hard to play against, definitely up for seeing how this goes. I think just having the option to climb over tables is definitely still worth the transformation, though.

**Comment by mybluecorners** [comment]
i would like it if it did a Throw on the monkey instead of teleporting them to the other side, but I still approve even if that doesnt get added

**Comment by MarkNais** [comment]
I very much like the idea of a throw! I'll do that.

File: code/obj/table.dm [hunk 1]
```diff
@@ -309,6 +309,8 @@
 				for (var/mob/N in AIviewers(usr, null))
 					if (N.client)
 						shake_camera(N, 4, 1, 0.5)
+		if(ismonkey(user))
```
Comment (ZeWaka) [inline code/obj/table.dm:312]: Couldn't you scope this inside the ishuman() above?

File: code/obj/table.dm [hunk 2]
```diff
@@ -354,6 +356,47 @@
...
 		..()
 
+	Bumped(atom/AM)
+		..()
+		if(!ismonkey(AM))
+			return
+		var/mob/living/carbon/human/M = AM
+		if(M.stat)
+			return
+		actions.start(new /datum/action/bar/icon/railing_jump/table_jump(AM, src), AM)
```
Comment (ZeWaka) [inline code/obj/table.dm:366]:
```suggestion
		actions.start(new /datum/action/bar/icon/railing_jump/table_jump(M, src), M)
```

File: code/obj/table.dm [hunk 3]
```diff
@@ -309,6 +309,8 @@
 				for (var/mob/N in AIviewers(usr, null))
 					if (N.client)
 						shake_camera(N, 4, 1, 0.5)
+			if(istype(user:mutantrace, /datum/mutantrace/monkey))
```
Comment (ZeWaka) [inline code/obj/table.dm:312]: `:`?

Comment (MarkNais) [reply code/obj/table.dm]: I don't know why I thought that was acceptable.
Memory tells me I copied it from somewhere, but I can't fine `user:mutantrace` anywhere else in the codebase

**Comment by MarkNais** [comment]
Colon removed. Still works great.
ZG8mQaYWMb

File: code/obj/table.dm [hunk 4]
```diff
@@ -309,6 +309,8 @@
 				for (var/mob/N in AIviewers(usr, null))
 					if (N.client)
 						shake_camera(N, 4, 1, 0.5)
+			if(istype(H.mutantrace, /datum/mutantrace/monkey))
```
Comment (ZeWaka) [inline code/obj/table.dm:312]:
```suggestion
			if(ismonkey(H))
```

File: code/obj/table.dm [hunk 5]
```diff
@@ -354,6 +356,47 @@
 			return src.desk_drawer.MouseDrop(over_object, src_location, over_location)
 		..()
 
+	Bumped(atom/AM)
+		..()
+		if(!ismonkey(AM))
+			return
+		var/mob/living/carbon/human/M = AM
+		if(M.stat)
```
Comment (ZeWaka) [inline code/obj/table.dm:364]:
```suggestion
		if(!isalive(M))
```

File: code/obj/table.dm [hunk 6]
```diff
@@ -354,6 +356,47 @@
...
+		actions.start(new /datum/action/bar/icon/railing_jump/table_jump(M, src), M)
+
+//Replacement for monkies walking through tables: They now parkour over them.
+//Note: Max count of tables traversable is 2 more than the iteration limit
+/datum/action/bar/icon/railing_jump/table_jump
+	getLandingLoc()
+		var/iteration = 0
+		var/turf/target = get_step(the_railing, ownerMob.dir)
+		var/obj/table/maybe_table = locate(/obj/table) in target
+		while(maybe_table && iteration < 5)
```
Comment (ZeWaka) [inline code/obj/table.dm:375]: magic number 5, var/const or #define it

File: code/obj/table.dm [hunk 7]
```diff
@@ -354,6 +356,47 @@
...
+		var/throw_speed = 0.5
+		if (istype(ownerMob, /mob/living/carbon/human))
+			var/mob/living/carbon/human/M = ownerMob
+			if (!(M.flags & TABLEPASS))
+				var/tables_traveled = duration/(1 SECONDS)
+				SPAWN_DBG((tables_traveled/throw_speed) DECI SECONDS)
+					M.flags &= !TABLEPASS		
+			M.flags ^= TABLEPASS
+
+		ownerMob.throw_at(jump_target, 7, throw_speed)
```
Comment (ZeWaka) [inline code/obj/table.dm:392]: ditto magic number

**Comment by pali6** [comment]
I like this, I'm okay with merging this when zewaka's minor concerns get addressed.

**Pull request merged**